dali/kernels/reduce/reduce_gpu_impl.cuh

```diff
@@ -0,0 +1,1058 @@
+      pow = 1;
+    int macroblocks = 1 << pow;
+    return macroblocks;
+  }
+
+  /**
+   * @brief Initializes a full reduction - per-sample or whole batch.
+   */
+  void InitReduceAll() {
+    const int nsamples = in_shape_.num_samples();
+    int64_t max_reduced_extent = 0;
+    for (int i = 0; i < nsamples; i++) {
+      int64_t extent = volume(in_shape_.tensor_shape_span(i));
+      if (extent > max_reduced_extent) {
+        max_reduced_extent = extent;
+      }
+    }
+
+    if (max_reduced_extent == 0) {
+      stages_.resize(1);
+      ReductionStage &stage = stages_.front();
+      stage.index = 0;
+      // stage 0 is never ReduceAll, because that would require contiguous input
+      stage.kind = reduce_batch_ ? ReductionKind::All : ReductionKind::Sample;
+      stage.shape.resize(stage.kind == ReductionKind::All ? 1 : nsamples);
+      for (auto &rs : stage.shape) {
+        rs.outer = 1;
+        rs.inner = 1;
+        rs.reduced_out = 1;
+        rs.reduced_in = 0;
+      }
+      stage.is_last = true;
+      return;
+    }
+
+    int log2max = ilog2(max_reduced_extent * (reduce_batch_ ? nsamples : 1));
+    int substages = div_ceil(log2max, 15);
+    if (substages == 1 && reduce_batch_ && nsamples > 1) {
+      // cannot reduce multiple samples in one stage due to contiguity requirement
+      substages++;
+    }
+
+    stages_.resize(substages);
+
+    for (int s = 0; s < substages; s++) {
+      ReductionStage &stage = stages_[s];
+      stage.index = s;
+      // stage 0 is never ReduceAll, because that would require contiguous input
+      stage.kind = s == 0 ? ReductionKind::Sample
+                          : reduce_batch_ ? ReductionKind::All : ReductionKind::Block;
+      stage.shape.resize(stage.kind == ReductionKind::All ? 1 : nsamples);
+    }
+
+    for (int i = 0; i < nsamples; i++) {
+      int64_t r = volume(in_shape_.tensor_shape_span(i));
+      auto &rs = stages_[0].shape[i];
+      rs.reduced_in = r;
+      rs.inner = rs.outer = 1;
+    }
+
+    for (int s = 0; s < substages; s++) {
+      int remaining_stages = substages - 1 - s;
+      ReductionStage &stage = stages_[s];
+      if (stage.kind == ReductionKind::All) {
+        // ReduceAll cannot happen at stage 0
+        assert(s > 0 || nsamples == 1);
+        int64_t r = 0;
+        if (s > 0) {
+          for (auto &rs : stages_[s-1].shape)
+            r += rs.reduced_out;
+        } else {
+          r = max_reduced_extent;
+        }
+        stage.shape[0].inner = stage.shape[0].outer = 1;
+        stage.shape[0].reduced_in = r;
+        r = CalcReducedExtent(r, remaining_stages);
+        stage.shape[0].reduced_out = r;
+      } else {
+        int64_t max_in = 0;
+        if (s > 0) {
+          for (int i = 0; i < nsamples;  i++) {
+            auto &rs = stage.shape[i];
+            int64_t r = stages_[s-1].shape[i].reduced_out;
+            rs.reduced_in = r;
+            rs.inner = rs.outer = 1;
```

Comment:
Why the reduced_in, inner and outer are set only in the `if (s > 0)` case?

Reply:
It's missing, but they're not used in ReduceAll/ReduceBlock anyway.